PhpdocOrderByValueFixer - Introduction

localheinz opened the pull request:
This PR

* [x] extracts a `PhpdocOrderByValueFixer`
* [x] deprecates the `PhpUnitOrderedCoversFixer`

localheinz commented on README.rst:
```diff
@@ -1398,9 +1398,19 @@ Choose from the list of available rules:
+  Configuration options:
+
+  - ``annotations`` (a subset of ``['covers', 'uses']``): list of annotations to
+    order, e.g. ``["covers"]``; defaults to ``[]``
```
Should we include `covers` by default, or should the default be an empty `array`?

SpacePossum replied:
I think `covers` as default is nice for the people migrating to the new fixer

localheinz commented on src/Fixer/PhpUnit/PhpUnitOrderedAnnotationsFixer.php:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    /**
+     * {@inheritdoc}
```
Do we really need these DocBlocks?

julienfalque replied:
We don't but I would leave it for consistency with the project. Maybe we'll revise this with #4500.

localheinz commented on README.rst:
```diff
@@ -1402,9 +1402,19 @@ Choose from the list of available rules:
+
+  Configuration options:
+
+  - ``annotations`` (a subset of ``['covers', 'uses']``): list of annotations to
```
Perhaps other annotations also make sense (for example, `@dataProvider`)?

SpacePossum replied:
maybe add these: `@group`, `@author`, `@dataProvider` ?

localheinz replied:
Crazy thing is that I heard from @epdenouden that some people write tests where the order in which data providers are configured matter.

😆

localheinz replied:
@SpacePossum 

Not sure about `@author` - would this make sense outside of the scope of a `phpunit/phpunit` test case?

SpacePossum replied:
well that is .. ehmm .... interesting XD    anyway... they can omit the option than : P

SpacePossum replied:
Im definitely fine with leaving `author` out, I was more typing out loud than pushing specs :)

localheinz replied:
@SpacePossum 

Had not been aware that `@author` is a thing as an annotation for `phpunit/phpunit`!

<img width="1104" alt="Screen Shot 2019-10-11 at 21 29 56" src="https://user-images.githubusercontent.com/605483/66679392-5f3e8000-ec6e-11e9-986c-f22a33f9472c.png">

localheinz replied:
Not a fan of using `@depends`, but does anyone know whether one can specify multiple `@depends`?

SpacePossum replied:
I never new it was an alias, TIL :D

SpacePossum replied:
https://phpunit.de/manual/6.5/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.examples.MultipleDependencies.php

some a sample of mulitple depends (not a big fan either btw)

localheinz replied:
@SpacePossum 

Added~

SpacePossum commented:
I would like to merge this one, so if anyone has more this please let me know otherwise I resolve the minors and continue :)

SpacePossum commented on tests/Fixer/PhpUnit/PhpUnitOrderedAnnotationsFixerTest.php:
```diff
@@ -0,0 +1,306 @@
+            ],
+        ];
+    }
+}
```
maybe we can have a test with multiple annotations configured as well?

localheinz commented on tests/Fixer/PhpUnit/PhpUnitOrderedAnnotationsFixerTest.php:
```diff
@@ -0,0 +1,954 @@
+                    class FooTest extends \PHPUnit_Framework_TestCase {}
+                ',
+            ],
+            'preserve positions if other docblock parts are present' => [
```
This test case currently fails.

🤔

SpacePossum replied:
I think we (you ;) ) should not need to trouble ourselves to much with `@author` annotations sticking to a piece of PHPDoc. 
For such details I think people can better refer to their VCS and only a list of authors per file can be handy. However, if they really want to keep the authors sticky to the rest of the text than they can remove the `@author` annotation from the configuration of the rule.

localheinz replied:
@SpacePossum 

Fixed it, had something messed up in my expectations!

👍

localheinz commented:
@julienfalque @SpacePossum 

Perhaps it makes sense to adjust the `PhpdocOrderFixer` to also take into account a range of other annotations?

I like

- having annotations in a specific order
- having annotations grouped (separated by an empty line)

What do you think?

SpacePossum commented:
I think that would be great :)
Do you want to incorporate the work done here in the new fixer into that one or keep these separated? I might be easier to have to prevent conflicts and make the configuration for the end user more easy... hmm... I might need to give it some more though, but it does sound good

localheinz commented:
Maybe there’s room for more fixers?

- group annotations
- sort annotations within a group
- sort annotation groups

Personally, I would use all of them!

localheinz commented:
Build failures are currently related to failing to download static analysis tools.

kubawerlos commented on README.rst:
```diff
@@ -1445,9 +1445,20 @@ Choose from the list of available rules:
+
+  Configuration options:
+
+  - ``annotations`` (a subset of ``['author', 'covers', 'dataProvider', 'depends',
```
What about `coversNothing` and `internal`? The more the merrier in this case, right?

keradus replied:
ping @localheinz

localheinz replied:
@kubawerlos 

Not sure, why would anyone have multiple

- `@coversNothing`
- `@internal`

annotations, perhaps even followed by content?

Should I add them nonetheless?

Perhaps a case for an additional fixer, remove duplicate annotations?

kubawerlos replied:
@localheinz oh, I was thinking the fixer will also order the groups of annotations :man_facepalming:

kubawerlos commented on src/Fixer/PhpUnit/PhpUnitOrderedAnnotationsFixer.php:
```diff
@@ -0,0 +1,184 @@
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        if (!\count($this->configuration['annotations'])) {
```
Boolish, why not `[] === $this->configuration['annotations']`?

keradus commented:
@localheinz , can you please rebase and handle newest code review ?

localheinz commented:
@keradus 

Apologies for the delay - will do!

localheinz commented:
@julienfalque @keradus @kubawerlos @SpacePossum 

I have

- rebased on top of `master`
- added handling of `@coversNothing`
- added handling of `@internal`

and I apologize for the delay.

If you find something else that needs to be addressed, I promise to sort it out quickly!

localheinz commented on README.rst:
```diff
@@ -1452,9 +1452,20 @@ Choose from the list of available rules:
+
+  Configuration options:
+
+  - ``annotations`` (a subset of ``['author', 'covers', 'coversNothing',
```
Just thinking, since this fixer handles annotations which are not bound to `phpunit` test cases, such as

- `@author`
- `@internal`

and since we already have `PhpdocOrderFixer`, perhaps this fixer here should be renamed to `PhpdocOrderByValueFixer`?

Here's what I think:

### Current Fixers

- `PhpDocOrderFixer` groups annotations by name (`@param`, `@throws`, `@return`) and inserts an empty line between them

### New Fixers

- `PhpDocOrderByValueFixer` sorts configured annotations by value
- `PhpDocOrderByNameFixer` sorts configured annotations by name, and optionally inserts a blank line between them

### Deprecated Fixers

* `PhpDocOrderFixer` can be deprecated and proxy to the `PhpDocOrderByNameFixer`

localheinz replied:
@julienfalque @keradus @kubawerlos @SpacePossum 

What do you think?

julienfalque replied:
That's exactly what I was thinking when reading the changes, 👍 for me. Though I think we can merge this PR (after renaming the new fixer) and work on the other fixers in another PR?

localheinz replied:
@julienfalque 

Sounds good!

Renamed the fixer!

julienfalque commented:
Thank you @localheinz.

localheinz commented:
Thank you, @julienfalque, @keradus, @kubawerlos, and @SpacePossum!

keradus commented:
kudos @localheinz
